File: src/main/scala/pl/iogreen/euler/Problem002.scala
package pl.iogreen.euler

import collection.mutable.ListBuffer

class Problem002 {

  def resolve(maxValue: Int): Int = {
    fibonacci(maxValue).filter(isEven(_)).sum
  }

  def fibonacci(maxValue: Int): List[Int] = {
    val foundFibonacci = ListBuffer[Int]()
    var nValue = 0
    var nPlusOneValue = 1

    while (nValue < maxValue) {
      val tempValue = nValue
      nValue = nextFibo(nValue, nPlusOneValue)
      nPlusOneValue = tempValue
      foundFibonacci.append(tempValue)
    }

    foundFibonacci.drop(2).toList
  }

  def isEven(n: Int): Boolean = {
    n % 2 == 0
  }

  def nextFibo(x: Int, y: Int): Int = {
    x + y
  }
}

File: kotlin/src/main/kotlin/pl/iogreen/euler/Problem002.kt
package pl.iogreen.euler.kotlin

public class Problem002 {

    fun solve(n: Int): Int {
        val zero = 0.toLong()

        var prev: Long = 1
        var curr: Long = 2
        var sum = 2

        while (true ) {
            val tmp = curr
            curr = curr + prev
            prev = tmp

            if (curr > n) {
                return sum
            } else if (curr % 2 == zero) {
                sum += curr
            }
        }
    }
}